File: Leetcode/BiweeklyContest149/RescheduleMeetingsforMaximumFreeTimeI.rs
impl Solution {
    pub fn max_free_time(event_time: i32, k: i32, start_time: Vec<i32>, end_time: Vec<i32>) -> i32 {
        let n = start_time.len();
        let k = k as usize;

        let mut e = vec![];
        for i in 0..n {
            e.push((start_time[i], end_time[i]));
        }
        e.push((0, 0));
        e.push((event_time, event_time));
        e.sort_unstable();
        // n + 2

        let mut ma = 0;
        // n + 1
        let mut slots = vec![];
        for w in e.windows(2) {
            let (e1, e2) = (w[0], w[1]);
            let free = e2.0 - e1.1;

            ma = ma.max(free);

            slots.push(free);
        }

        let mut free = 0;
        let k = k + 1;
        for i in 0..k {
            free += slots[i];
        }
        ma = ma.max(free);

        for i in k..slots.len() {
            free -= slots[i - k];
            free += slots[i];
            ma = ma.max(free);
        }

        ma
    }
}

File: Leetcode/RescheduleMeetingsforMaximumFreeTimeI.cpp
// https://leetcode.com/problems/reschedule-meetings-for-maximum-free-time-i

class Solution {
   public:
    int maxFreeTime(int eventTime,
                    int k,
                    vector<int>& startTime,
                    vector<int>& endTime) {
        size_t n = startTime.size();

        vector<int> startTimeExt(n + 2, 0);
        vector<int> endTimeExt(n + 2, 0);

        for (size_t i = 0; i < n; ++i) {
            startTimeExt[i + 1] = startTime[i];
            endTimeExt[i + 1] = endTime[i];
        }

        startTimeExt[n + 1] = eventTime;
        endTimeExt[n + 1] = eventTime;

        vector<int> pref(n + 3, 0);
        for (size_t i = 0; i <= n + 1; ++i) {
            pref[i + 1] += pref[i] + (endTimeExt[i] - startTimeExt[i]);
        }

        int ans = 0;
        for (size_t i = k; i < n + 1; ++i) {
            ans = max(ans, startTimeExt[i + 1] - endTimeExt[i - k] -
                               (pref[i + 1] - pref[i - k + 1]));
        }

        return ans;
    }
};